fix: refresh auth token every 7 days

sentientwaffle opened the pull request:
* Fix https://github.com/interledgerjs/ilp-plugin-bells/issues/129
* Fix a typo in `Factory#_create`.

Depends on https://github.com/interledgerjs/ilp-plugin-bells/pull/132 and https://github.com/interledgerjs/five-bells-ledger/pull/425 and https://github.com/interledgerjs/five-bells-ledger/pull/429

Commit:
feat: adds support for token authentication

sentientwaffle committed:
fix: refresh auth token every 7 days

* Fix https://github.com/interledgerjs/ilp-plugin-bells/issues/129
* Fix a typo in `Factory#_create`.

Depends on https://github.com/interledgerjs/ilp-plugin-bells/pull/132 and https://github.com/interledgerjs/five-bells-ledger/pull/425

sharafian commented on src/lib/plugin.js:
```diff
@@ -208,7 +209,11 @@ class FiveBellsLedger extends EventEmitter2 {
-      return new WebSocket(wsUri)
+      return new WebSocket(wsUri, null, {
+        headers: {
+          authorization: 'Bearer ' + options.authToken
```
I thought the reason we were using the token in the URI was because the browser doesn't support these websocket headers. Won't this break browser support?

sentientwaffle replied:
That change is part of https://github.com/interledgerjs/ilp-plugin-bells/pull/132, which this PR is based on. @dappelt can you weigh in?

dappelt replied:
Propably it won't work in the browser. I will investigate a workaround.

dappelt replied:
This does not work out of the box in a browser. Let's stick with appending the token to the URL for now.

sharafian commented on src/lib/plugin.js:
```diff
@@ -25,6 +25,7 @@ const defaultConnectTimeout = 60000
 const wsReconnectDelayMin = 10
 const wsReconnectDelayMax = 500
 const defaultMessageTimeout = 5000
+const authTokenMaxAge = 7 * 24 * 60 * 60 * 1000 // one week
```
Will the websocket stop working if the token becomes too old? If so, we might want to set this age to less than a week, so that it's refreshed before it expires.

sentientwaffle replied:
I don't think so. The token is only checked once, when the websocket first connects.

dappelt replied:
> Will the websocket stop working if the token becomes too old?

No, it should not stop working. Authentication is done before the websocket connection is established.

emschwartz commented:
I think it might be better to have the ledger actually tell you how long the token is valid for so the plugin knows when it needs to reconnect. Otherwise the actual time we implement would have to be considered part of the ledger protocol (what if someone sets a shorter time than what the plugin assumes?)

sentientwaffle committed:
fix: use /auth_token token_max_age

Related to https://github.com/interledgerjs/ilp-plugin-bells/pull/136#issuecomment-307850559

Depends on https://github.com/interledgerjs/five-bells-ledger/pull/429

Commit:
feat: Send auth token in the query string

dappelt reviewed (commented):
LGTM. I will merge this as part of https://github.com/interledgerjs/ilp-plugin-bells/pull/137
